core/src/main/scala/org/apache/spark/launcher/LauncherBackend.scala

```diff
@@ -71,6 +110,9 @@ private[spark] abstract class LauncherBackend {
     if (connection != null && lastState != state) {
       connection.send(new SetState(state))
       lastState = state
+      if(!_isConnected && stopFlag) {
```

Comment:
ok perhaps we can move this into the close() call in the BackendConnection then so we don't have to do if on every setState.